core-rust/state-manager/src/mempool/pending_transaction_result_cache.rs

```diff
@@ -513,134 +549,202 @@ const NON_REJECTION_RECALCULATION_DELAY: Duration = Duration::from_secs(120);
 const MAX_RECALCULATION_DELAY: Duration = Duration::from_secs(1000);
 
 pub struct PendingTransactionResultCache {
-    mempool: Arc<RwLock<PriorityMempool>>,
-    pending_transaction_records: LruCache<NotarizedTransactionHash, PendingTransactionRecord>,
+    pending_transaction_records: LruCache<
+        NotarizedTransactionHash,
+        (
+            PendingTransactionRecord,
+            TransactionIntentHash,
+            Vec<SubintentHash>,
+        ),
+    >,
+    // INVARIANT: The `intent_lookup` and `subintent_lookup` are kept exactly in sync with
```

Comment:
Is mempool transaction adding single-threaded? The "sync" logic doesn't look thread-safe - there is a race condition where the reverse lookup goes out-of-sync with pending transaction records.

Reply:
You're right. The `&mut` references are obtained from `write()` on the RW Lock, not some unsafe transmutation, which prevents unsynchronized modification to the cache data structure.